### PR TITLE
Let log_artifact log directories

### PR DESCRIPTION
**Pull request by apurva-koti**
## What changes are proposed in this pull request?
 
(Not ready for merge - making a PR so we can discuss this change)
Users must use `log_artifact` for files and `log_artifacts` for directories. Until recently, documentation incorrectly stated the former could be used for both. This PR brings the two under `log_artifact` and deprecates `log_artifacts`.
 
## How is this patch tested?

Manual, unit
 
### Is this a user-facing change? 

- [ ] No. You can skip the rest of this section.
- [X] Yes. Give a description of this change to be included in the release notes for MLflow users.
 
`log_artifacts` is deprecated. Use `log_artifact`, which can now log both files and directories.
 
### What component(s) does this PR affect?
 
- [ ] UI
- [ ] CLI 
- [ ] API 
- [ ] REST-API 
- [ ] Examples 
- [ ] Docs
- [X] Tracking
- [ ] Projects 
- [ ] Artifacts 
- [ ] Models 
- [ ] Scoring 
- [ ] Serving
- [ ] R
- [ ] Java
- [ ] Python

### How should the PR be classified in the release notes? Choose one:
 
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [X] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/tracking/fluent.py [hunk 1]
```diff
@@ -256,9 +257,13 @@ def log_artifact(local_path, artifact_path=None):
     :param artifact_path: If provided, the directory in ``artifact_uri`` to write to.
     """
     run_id = _get_or_start_run().info.run_id
+    if os.path.isdir(local_path):
+        MlflowClient().log_artifacts(run_id, local_path, artifact_path)
```
Comment (smurching) [inline mlflow/tracking/fluent.py]: Should probably put the `log_artifact` call below in an `else` :P

Comment (smurching) [reply mlflow/tracking/fluent.py]: Or actually can't we just always call `MlflowClient.log_artifact()` now that it's been fixed to accept a file or directory?

File: mlflow/tracking/client.py [hunk 2]
```diff
@@ -218,7 +219,10 @@ def log_artifact(self, run_id, local_path, artifact_path=None):
         """
         run = self.get_run(run_id)
         artifact_repo = get_artifact_repository(run.info.artifact_uri)
-        artifact_repo.log_artifact(local_path, artifact_path)
+        if os.path.isdir(local_path):
+            artifact_repo.log_artifacts(local_path, artifact_path)
+        else:
+            artifact_repo.log_artifact(local_path, artifact_path)
 
     def log_artifacts(self, run_id, local_dir, artifact_path=None):
```
Comment (smurching) [inline mlflow/tracking/client.py]: Should we deprecate this as well?

**Review by smurching — COMMENTED** [review]
Mostly looks good, we should also add tests that pass directories to mlflow.log_artifact, MlflowClient.log_artifact, Java's MlflowClient.logArtifact APIs and verify that things work as expected.

File: mlflow/tracking/fluent.py (hunk 1 above)
Comment (apurva-koti) [reply mlflow/tracking/fluent.py]: Good call, thanks

File: mlflow/tracking/fluent.py [hunk 3]
```diff
@@ -257,15 +258,16 @@ def set_tags(tags):
...
     """
-    Log a local file as an artifact of the currently active run.
+    Log a local file or directory as an artifact of the currently active run.
 
     :param local_path: Path to the file to write.
     :param artifact_path: If provided, the directory in ``artifact_uri`` to write to.
     """
     run_id = _get_or_start_run().info.run_id
-    MlflowClient().log_artifact(run_id, local_path, artifact_path)
+    MlflowClient().log_artifacts(run_id, local_path, artifact_path)
```
Comment (dbczumar) [inline mlflow/tracking/fluent.py]: Looks like we're calling a deprecated method here....?

File: mlflow/tracking/client.py [hunk 4]
```diff
@@ -218,15 +220,19 @@ def log_batch(self, run_id, metrics=(), params=(), tags=()):
 
     def log_artifact(self, run_id, local_path, artifact_path=None):
         """
-        Write a local file to the remote ``artifact_uri``.
+        Write a local file or directory to the remote ``artifact_uri``.
```
Comment (dbczumar) [inline mlflow/tracking/client.py]: After testing, I've confirmed that `mlflow.log_artifacts` has the same semantics as `mlflow.log_artifact` for a directory input: MLflow logs the *contents* of the directory; it does not create the outer directory in the artifact store.

E.g., consider the following directory structure:
```
    outer_dir
    ├── inner_dir
    |   ├── inner_file.txt
    ├── file.txt
```

Calling `mlflow.log_artifact(s)("output_dir")` produces the following artifact contents:
```
    ├── inner_dir
    |   ├── inner_file.txt
    ├── file.txt
```
`outer_dir` is omitted.

Can we make this behavior clear in the documentation?

File: mlflow/tracking/fluent.py [hunk 5]
```diff
@@ -257,7 +258,7 @@ def set_tags(tags):
 
 def log_artifact(local_path, artifact_path=None):
```
Comment (dbczumar) [inline mlflow/tracking/fluent.py]: We should add tests for the behavior of this method when it's invoked on a directory.

File: mlflow/tracking/client.py [hunk 6]
```diff
@@ -218,15 +220,19 @@ def log_batch(self, run_id, metrics=(), params=(), tags=()):
 
     def log_artifact(self, run_id, local_path, artifact_path=None):
```
Comment (dbczumar) [inline mlflow/tracking/client.py]: We should add tests for the behavior of this method when it's invoked on a directory.

File: mlflow/tracking/client.py (hunk 4 above)
Comment (apurva-koti) [reply mlflow/tracking/client.py]: I actually think it's more sensible to create the outer directory. Logging a file makes that file visible as an artifact, so logging a directory should make that directory visible as is. 
@smurching @mateiz  thoughts?

**Comment by apurva-koti** [comment]
This PR will probably remain a WIP for a few days due to the high number of uses of `log_artifacts` in MLflow code that need to be removed.

File: mlflow/tracking/fluent.py (hunk 5 above)
Comment (apurva-koti) [reply mlflow/tracking/fluent.py]: Would it suffice to change existing tests on `log_artifacts` to use `log_artifact`?

**Comment by apurva-koti** [comment]
Offline discussion with @smurching - we'll scope this PR just to add directory support to `log_artifact` and handle the deprecation and code change in another PR.

**Pull request closed without merging**